From 43eabd8820e391dcfdc0a9b79e088b7d59704967 Mon Sep 17 00:00:00 2001
From: Mingcong Bai <jeffbai@aosc.io>
Date: Mon, 27 Jan 2025 04:56:34 +0800
Subject: [PATCH 3/9] AOSCOS: LOONGSON3: 70kernel-modules: minimise default
 kernel drivers

Make sure that only drivers that were likely needed at early-boot were
included as part of the initramfs.

Saves about 5MiB post-XZ.

Signed-off-by: Mingcong Bai <jeffbai@aosc.io>
Signed-off-by: Kaiyang Wu <origincode@aosc.io>
---
 modules.d/70kernel-modules/module-setup.sh | 93 ++++++++++++++--------
 1 file changed, 61 insertions(+), 32 deletions(-)

diff --git a/modules.d/70kernel-modules/module-setup.sh b/modules.d/70kernel-modules/module-setup.sh
index 716a5b5e..f516e32b 100755
--- a/modules.d/70kernel-modules/module-setup.sh
+++ b/modules.d/70kernel-modules/module-setup.sh
@@ -25,12 +25,22 @@ installkernel() {
     }
 
     install_block_modules() {
-        hostonly=$(optional_hostonly) instmods \
-            scsi_dh_rdac scsi_dh_emc scsi_dh_alua \
-            =drivers/usb/storage \
-            =ide nvme vmd \
-            virtio_blk virtio_scsi \
-            =drivers/ufs
+        if [[ ${DRACUT_ARCH:-$(uname -m)} != mips64 ]]; then
+            hostonly=$(optional_hostonly) instmods \
+                scsi_dh_rdac scsi_dh_emc scsi_dh_alua \
+                =drivers/usb/storage \
+                =ide nvme vmd \
+                virtio_blk virtio_scsi \
+                =drivers/ufs
+        else
+            # MIPS64: Minimise kernel modules to save initramfs size,
+            # taking into account USB mass storage, IDE/ATA, NVMe, and
+            # virtualised storage (virtio).
+            hostonly=$(optional_hostonly) instmods \
+                =drivers/usb/storage \
+                =ide nvme \
+                virtio_blk virtio_scsi
+        fi
 
         dracut_instmods -o -s "${_blockfuncs}" "=drivers"
     }
@@ -39,32 +49,51 @@ installkernel() {
         hostonly='' instmods \
             hid_generic unix
 
-        # Some xHCI PCI controllers are handled by sibling glue drivers
-        # instead of plain xhci-pci. Include them so USB is available
-        # during initrd when those drivers are built as modules.
-        hostonly=$(optional_hostonly) instmods \
-            ehci-hcd ehci-pci ehci-platform \
-            ohci-hcd ohci-pci \
-            uhci-hcd \
-            usbhid \
-            xhci-hcd xhci-pci xhci-pci-prom21 xhci-pci-renesas xhci-plat-hcd \
-            "=drivers/hid" \
-            "=drivers/tty/serial" \
-            "=drivers/input/serio" \
-            "=drivers/input/keyboard" \
-            "=drivers/mmc/core" \
-            "=drivers/mmc/host" \
-            "=drivers/pci/host" \
-            "=drivers/pci/controller" \
-            "=drivers/pinctrl" \
-            "=drivers/usb/typec" \
-            "=drivers/watchdog"
-
-        hostonly=$(optional_hostonly) instmods \
-            yenta_socket intel_lpss_pci spi_pxa2xx_platform \
-            atkbd i8042 firewire-ohci hv-vmbus \
-            virtio virtio_ring virtio_pci pci_hyperv \
-            surface_aggregator_registry psmouse
+        if [[ ${DRACUT_ARCH:-$(uname -m)} != mips64 ]]; then
+            # Some xHCI PCI controllers are handled by sibling glue drivers
+            # instead of plain xhci-pci. Include them so USB is available
+            # during initrd when those drivers are built as modules.
+            hostonly=$(optional_hostonly) instmods \
+                ehci-hcd ehci-pci ehci-platform \
+                ohci-hcd ohci-pci \
+                uhci-hcd \
+                usbhid \
+                xhci-hcd xhci-pci xhci-pci-prom21 xhci-pci-renesas xhci-plat-hcd \
+                "=drivers/hid" \
+                "=drivers/tty/serial" \
+                "=drivers/input/serio" \
+                "=drivers/input/keyboard" \
+                "=drivers/mmc/core" \
+                "=drivers/mmc/host" \
+                "=drivers/pci/host" \
+                "=drivers/pci/controller" \
+                "=drivers/pinctrl" \
+                "=drivers/usb/typec" \
+                "=drivers/watchdog"
+        else
+            # MIPS64: Minimise kernel modules to save initramfs size,
+            # taking into account generic USB HID drivers and USB hosts.
+            hostonly=$(optional_hostonly) instmods \
+                ehci-hcd ehci-pci ehci-platform \
+                ohci-hcd ohci-pci \
+                uhci-hcd \
+                usbhid \
+                xhci-hcd xhci-pci xhci-plat-hcd
+        fi
+
+        if [[ ${DRACUT_ARCH:-$(uname -m)} != mips64 ]]; then
+            hostonly=$(optional_hostonly) instmods \
+                yenta_socket intel_lpss_pci spi_pxa2xx_platform \
+                atkbd i8042 firewire-ohci hv-vmbus \
+                virtio virtio_ring virtio_pci pci_hyperv \
+                surface_aggregator_registry psmouse
+        else
+            # MIPS64: Minimise kernel modules to save initramfs size,
+            # taking into account AT keyboards, i8042, and virtio devices.
+            hostonly=$(optional_hostonly) instmods \
+                atkbd i8042 \
+                virtio virtio_ring virtio_pci
+        fi
 
         if [[ ${DRACUT_ARCH} == arm* || ${DRACUT_ARCH} == aarch64 || ${DRACUT_ARCH} == riscv* ]]; then
             # arm/aarch64 specific modules
-- 
2.55.0

